pkg/scheduler/queue/queue.go

```diff
@@ -96,26 +98,29 @@ func (q *RequestQueue) EnqueueRequest(userID string, req Request, maxQueriers fl
 	}
 
 	shardSize := util.DynamicShardSize(maxQueriers, len(q.queues.queriers))
-	queue := q.queues.getOrAddQueue(userID, shardSize)
+	queue := q.queues.getOrAddQueue(userID, shardSize, req.IsHighPriority())
+	maxOutstandingRequests := q.queues.limits.MaxOutstandingPerTenant(userID)
+
 	if queue == nil {
 		// This can only happen if userID is "".
 		return errors.New("no queue found")
 	}
 
 	q.totalRequests.WithLabelValues(userID).Inc()
-	select {
-	case queue <- req:
-		q.queueLength.WithLabelValues(userID).Inc()
-		q.cond.Broadcast()
-		// Call this function while holding a lock. This guarantees that no querier can fetch the request before function returns.
-		if successFn != nil {
-			successFn()
-		}
-		return nil
-	default:
+
+	if q.queues.getTotalQueueSize(userID) >= maxOutstandingRequests {
```

Comment:
This means, a high priority request might still be discarded if there are already plenty of requests in the queue

Reply:
Yes, that's the implemented behavior. Are you suggesting to always accept high priority queries instead?